File: solutions/python/structure/fixed/StaticGraphADT.py
class StaticGraphADT:
    def __init__(self, max_vertices):
        self.max_vertices = max_vertices
        self.matrix = [[None for _ in range(max_vertices)] for _ in range(max_vertices)]
        self.vertices = set()

    def get_vertices(self):
        return self.vertices

    def add_vertex(self, vertex):
        if 0 <= vertex < self.max_vertices:
            self.vertices.add(vertex)

    def remove_vertex(self, vertex):
        if vertex in self.vertices:
            self.vertices.remove(vertex)
            for i in range(self.max_vertices):
                self.matrix[vertex][i] = None
                self.matrix[i][vertex] = None

    def add_edge(self, v1, v2, weight):
        if v1 in self.vertices and v2 in self.vertices:
            self.matrix[v1][v2] = weight

    def remove_edge(self, v1, v2):
        if v1 in self.vertices and v2 in self.vertices:
            self.matrix[v1][v2] = None

    def exists_edge(self, v1, v2):
        return self.matrix[v1][v2] is not None

    def edge_weight(self, v1, v2):
        if self.exists_edge(v1, v2):
            return self.matrix[v1][v2]
        raise ValueError("No existe la arista")

    def is_empty(self):
        return len(self.vertices) == 0

File: solutions/javascript/src/structure/fixed/StaticGraphADT.js
class StaticGraphADT {
  constructor(maxVertices = 10) {
    this.maxVertices = maxVertices;
    this.vertices = new Array(maxVertices).fill(null); // posiciones ocupadas por vértices o null
    this.adjMatrix = Array.from({ length: maxVertices }, () =>
      new Array(maxVertices).fill(0)
    );
  }

  findIndex(vertex) {
    return this.vertices.indexOf(vertex);
  }

  addVertx(vertex) {
    if (this.findIndex(vertex) !== -1) return;
    const index = this.vertices.indexOf(null);
    if (index !== -1) this.vertices[index] = vertex;
  }

  removeVertx(vertex) {
    const idx = this.findIndex(vertex);
    if (idx === -1) return;
    this.vertices[idx] = null;
    for (let i = 0; i < this.maxVertices; i++) {
      this.adjMatrix[idx][i] = 0;
      this.adjMatrix[i][idx] = 0;
    }
  }

  addEdge(from, to, weight) {
    const i = this.findIndex(from);
    const j = this.findIndex(to);
    if (i !== -1 && j !== -1) this.adjMatrix[i][j] = weight;
  }

  removeEdge(from, to) {
    const i = this.findIndex(from);
    const j = this.findIndex(to);
    if (i !== -1 && j !== -1) this.adjMatrix[i][j] = 0;
  }

  existsEdge(from, to) {
    const i = this.findIndex(from);
    const j = this.findIndex(to);
    return i !== -1 && j !== -1 && this.adjMatrix[i][j] !== 0;
  }

  edgeWeight(from, to) {
    const i = this.findIndex(from);
    const j = this.findIndex(to);
    if (i !== -1 && j !== -1) return this.adjMatrix[i][j];
    return null;
  }

  getVertxs() {
    const set = new SetADT();
    for (const v of this.vertices) if (v !== null) set.add(v);
    return set;
  }

  isEmpty() {
    return this.vertices.every(v => v === null);
  }
}

export default StaticGraphADT;